### ISSUE TITLE
Discussion: testing fit with log-normal distribution

### ISSUE THREAD

**Issue by gvwilson** [merely-useful/merely-useful.github.io#288]
1.  We want to add a script `testfit.py` that tests how well data fits the "expected" distribution according to Zipf's Law.
2.  We don't really know what's expected: should it be log-normal?

-   Thread: https://twitter.com/RobertDotKern/status/1198389738752892928
-   Jake Vanderplas article: https://jakevdp.github.io/blog/2017/11/09/exploring-line-lengths-in-python-packages/

**Comment by cwickham** [comment]
**Log Normal v.s Power Law**

* I didn't find any convincing mechanisms that would suggest word frequencies should be Log Normal.  In contrast, there is a reasonable (but unrealistic) mechanism for why word frequencies might be power law distributed.

* When the probability density function is plotted on a log-log scale:
    * it will appear **linear** (at least in the tails), for a power law distributed variable, and
    * it will appear **quadratic**, for a  Log Normally distributed variable.
    
    Therefore, @newman2005power suggests that since a quadratic can be well approximated by a line on small regions of the scale, a log normally distributed variable may well appear power law distributed.  You can probably make the reverse argument, that a line can be well fit by a quadratic (with its extra parameter), and as such, a log normal may well fit as well or better to data that is in fact power law distributed.

* A log normal is a distribution for continuous random variables, but one might argue word frequencies, since they are counts, are discrete.  *This may not matter much for large texts*.

**Recommendations**

* I suggest we focus on fitting the power law, and don't fit a Log Normal at all.  

* Following [Moreno-Sánchez, Font-Clos, and Corral
(2016)](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0147073) we fit the discrete form with the power law on the survival function with a single parameter: the exponent in the power function (the minimum value of frequency for which the power law applies is assumed to be 1).  

* Fit using maximum likelihood following recommendations of [Clauset, Shalizi, and Newman (2009)](https://arxiv.org/abs/0706.1062).

**Questions**

* Do we want to do any goodness of fit testing?  I.e. we can estimate the exponent, but do we want to ask, is a power law a good fit?  From `automate.Rmd` it looks like the answer is, yes.  This will be more involved code: the recommended approach involves a Monte Carlo simulation to get a p-value.

## Notes

I've have notes at https://github.com/cwickham/merely-useful.github.io/blob/fit-zipfs/zipfs-law/zipfs-law.md Mostly trying to get my head around all this, but may be useful when putting together the code for `testfit.py`(`.r`).

**Comment by DamienIrving** [comment]
Thanks, @cwickham. That rank frequency plot with fitted power law looks great!

**Comment by gvwilson** [comment]
:+1:

**Comment by DamienIrving** [comment]
@cwickham - Can I check my understanding with you?

The typical quote about Zipfs Law is as follows: "the most frequent word will occur approximately twice as often as the second most frequent word, three times as often as the third most frequent word, etc"

As you say in your document, that can be expressed mathematically as _f_ (frequency) is proportional to 1 / _r_ (rank).

A more general form adds a parameter alpha for the exponent of the power law (i.e. 1 / r^alpha).

You analysis of Moby Dick suggests that alpha is about 1.9, which means a better statement would be that "the most frequent word will occur approximately four times as often as the second most frequent word, nine times as often as the third most frequent word, etc"

Is that correct?

**Comment by cwickham** [comment]
@DamienIrving   Ah, there's a bit of an inconsistency in the notation in my writeup.  The quote above indicates a relationship something like "number of occurrences of word" &#8733; 1/"rank of word"<sup>&alpha;</sup>, with &alpha; = 1.  The parameter I estimated isn't actually &alpha;, let's call it &beta;, where &beta; = 1 + 1/&alpha;.  So, &beta; = 1.9, corresponds to &alpha; = 1.1, which is perhaps closer to the original quote than it initially appeared.

The inconsistency arises because that initial quote is talking about "how often a word occurs" as a function of rank, i.e. the values that exhibit the power law are word counts and we have as many counts as we have unique words. In the statistical setup we talk about "how many words have the same word count" as a function of rank, i.e. the values that exhibit the power law are counts of word counts and we have as many counts as we have unique word frequencies.  Sources often use &alpha; for the exponent in both settings, but they aren't the same value.  I think I'll try to clarify this in my writeup.

The plan sounds great.

**Comment by cwickham** [comment]
EDIT: Actually the inconsistency arises in a much more easily seen way.  In the statistical approach, we assume a power law on the survival function, S(x) &prop; x<sup>(1 - &beta;)</sup>, where x is the number of times a word occurs.  The typical quote, puts a power law on x as a function of rank, r, x &prop; r<sup>-&alpha;</sup>.  The rank, r, is a linear transform of survival so inverting this relationship gives r = S(x) &prop; x<sup>-1/&alpha;</sup> then setting exponents equal, -1/&alpha; = (1 - &beta;),  leads to &beta; = 1 + 1/&alpha;.

**Comment by DamienIrving** [comment]
@cwickham That makes sense - thanks!

Our statement for Moby Dick would therefore be "the most frequent word will occur approximately 2<sup>1.1</sup> = 2.14 times as often as the second most frequent word, 3<sup>1.1</sup> = 3.35 times as often as the third most frequent word, etc", which is pretty close to the original quote.

I now just need to re-write your process for estimating beta in Python. i.e. this bit:
```
# negative log likelihood l(beta) = -sum(log(f(x)))
nlog_likelihood <- function(beta, x){
  - sum(log((1/x)^(beta - 1) - (1/(x + 1))^(beta - 1)))
}

# Optimizer settings from @moreno2016large
mle <- optim(1.5, nlog_likelihood, x = moby$count, 
  lower = 1, upper = 4, 
  hessian = TRUE, method = "Brent")

beta_hat <- mle$par
# Standard error from ML theory
se <- sqrt(1/mle$hessian)[1, 1]

# Asymptotic 95% CI
ci <- beta_hat + c(-1, 1) * 1.96 * se
```
Have you done this kind of thing in Python before by any chance? It looks like I could use [`scipy.optimize.brent`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.brent.html)?

**Comment by cwickham** [comment]
@DamienIrving Nope...sorry no experience doing this Python.  But, yes, looks like [scipy.optimize.brent](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.brent.html) should work.   It's also a minimiser, so `nlog_likelihood()` can be a direct translation to Python and then `xmin` in the `scipy.optimize.brent` result will be the estimate of &beta;.  

Doesn't look like `scipy.optimize.brent` returns the Hessian which is needed if you want to put a confidence interval on the estimate.  It isn't clear to me if [`scipy.optimize.minimize_scalar(method = "Brent")`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize_scalar.html#scipy.optimize.minimize_scalar) is just calling `scipy.optimize.brent` or not, but in general it looks like that might return a Hessian in its [result](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html#scipy.optimize.OptimizeResult).

**Comment by DamienIrving** [comment]
I've now updated [`plotcounts.py`](https://github.com/DamienIrving/merely-useful.github.io/blob/zipfs_code/zipfs-law/bin/plotcounts.py) so that it uses `scipy.optimize.minimize_scalar` to estimate &beta; and plot the fitted power law curve.

This is what it looks like for Moby Dick (&alpha; = 1.093, &beta; = 1.915). (`minimize_scalar` didn't give the hessian so I couldn't estimate the confidence interval, but I'm not too fussed about that.)
![moby_dick](https://user-images.githubusercontent.com/2062210/71637896-7fe20900-2ca4-11ea-952f-0fe2995bbd59.png)

**Comment by joelostblom** [comment]
@cwickham @DamienIrving Minor comment: In Moreno-Sanchez_2016, the range/bounds is defined as 1 < β ≤ 4. In our Python implementation we're using `scipy.optimize.minimize_scalar`, which is inclusive for both the lower and the upper bond. This is a problem because

```python
(1/counts)**(beta - 1) - (1/(counts + 1))**(beta - 1)
```

evaluates to 0 when β=1, so taking the log of that raises a warning: `RuntimeWarning: divide by zero encountered in log`, which might confuse learners. I opened PR #333 to solve this by setting a slightly bigger number as the lower bond, and wanted to check with you if that makes sense (not sure if there is a more elegant solution). This change did not not impact the alpha and beta values on the Moby Dick text (same as damieans above)

I am note sure if this is a problem in the R implementation as well or if it is solved by `optim` having a param for the initial value (which we set to 1.5 above).
